private/bufpkg/bufplugin/bufpluginconfig/bufpluginconfig.go

```diff
@@ -0,0 +1,205 @@
+//
+// Only one field will be set.
+type RuntimeConfig struct {
+	Go      *GoRuntimeConfig
+	NPM     *NPMRuntimeConfig
+	Archive *ArchiveRuntimeConfig
+}
+
+// GoRuntimeConfig is the runtime configuration for a Go plugin.
+type GoRuntimeConfig struct {
+	MinVersion string
+	Deps       []string
+}
+
+// NPMRuntimeConfig is the runtime configuration for a JavaScript NPM plugin.
+type NPMRuntimeConfig struct {
+	Deps []string
+}
+
+// ArchiveRuntimeConfig is the runtime configuration for a plugin that can be downloaded
+// as an archive instead of a language-specific registry.
+type ArchiveRuntimeConfig struct {
+	Deps []string
+}
+
+// GetConfigForBucket gets the Config for the YAML data at ConfigFilePath.
+//
+// If the data is of length 0, returns the default config.
+func GetConfigForBucket(ctx context.Context, readBucket storage.ReadBucket) (*Config, error) {
+	return getConfigForBucket(ctx, readBucket)
+}
+
+// GetConfigForData gets the Config for the given JSON or YAML data.
+//
+// If the data is of length 0, returns the default config.
+func GetConfigForData(ctx context.Context, data []byte) (*Config, error) {
+	return getConfigForData(ctx, data)
+}
+
+// ExistingConfigFilePath checks if a configuration file exists, and if so, returns the path
+// within the ReadBucket of this configuration file.
+//
+// Returns empty string and no error if no configuration file exists.
+func ExistingConfigFilePath(ctx context.Context, readBucket storage.ReadBucket) (string, error) {
+	for _, configFilePath := range AllConfigFilePaths {
+		exists, err := storage.Exists(ctx, readBucket, configFilePath)
+		if err != nil {
+			return "", err
+		}
+		if exists {
+			return configFilePath, nil
+		}
+	}
+	return "", nil
+}
+
+// ParseConfig parses the file at the given path as a Config.
+func ParseConfig(config string) (*Config, error) {
+	var data []byte
+	var err error
+	switch filepath.Ext(config) {
+	case ".json", ".yaml", ".yml":
+		data, err = os.ReadFile(config)
+		if err != nil {
+			return nil, fmt.Errorf("could not read file: %w", err)
+		}
+	default:
+		return nil, fmt.Errorf("invalid extension %s, must be .json, .yaml or .yml", filepath.Ext(config))
+	}
+	var externalConfig ExternalConfig
+	if err := encoding.UnmarshalJSONOrYAMLStrict(data, &externalConfig); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal plugin config: %w", err)
+	}
+	switch externalConfig.Version {
+	case V1Version:
+		return newConfig(externalConfig)
+	}
+	return nil, fmt.Errorf("invalid plugin configuration version: must be one of %v", AllConfigFilePaths)
+}
+
+// ExternalConfig represents the on-disk representation
+// of the plugin configuration at version v1.
+type ExternalConfig struct {
+	Version string                `json:"version,omitempty" yaml:"version,omitempty"`
+	Name    string                `json:"name,omitempty" yaml:"name,omitempty"`
+	Opts    []string              `json:"opts,omitempty" yaml:"opts,omitempty"`
```

Comment:
Maybe just `options`? I don't know

Reply:
I guess `opts` is consistent with `deps, min_version`

Reply:
Yea, I'd like it to be consistent with the `buf.gen.yaml`, too.